human_aware_rl/rllib/rllib.py

```diff
@@ -0,0 +1,600 @@
+from ray.rllib.agents.ppo.ppo import PPOTrainer
+from ray.rllib.models import ModelCatalog
+from human_aware_rl.rllib.utils import softmax, get_base_env, get_mlp, get_required_arguments, iterable_equal
+from datetime import datetime
+import tensorflow as tf
+import inspect
+import ray
+import tempfile
+import gym
+import numpy as np
+import os, pickle, copy
+
+action_space = gym.spaces.Discrete(len(Action.ALL_ACTIONS))
+obs_space = gym.spaces.Discrete(len(Action.ALL_ACTIONS))
+timestr = datetime.today().strftime("%Y-%m-%d_%H-%M-%S")
+
+
+class RlLibAgent(Agent):
+    """ 
+    Class for wrapping a trained RLLib Policy object into an Overcooked compatible Agent
+    """
+    ### TODO: Figure out how to reset rnn_state at start of episode
+    def __init__(self, policy, agent_index, featurize_fn):
+        self.policy = policy
+        self.agent_index = agent_index
+        self.featurize = featurize_fn
+
+        # Get initial rnn states and add batch dimension to each
+        if hasattr(policy.model, 'get_initial_state'):
+            self.rnn_state = [np.expand_dims(state, axis=0) for state in policy.model.get_initial_state()]
+        else:
+            self.rnn_state = []
+
+    def action_probabilities(self, state):
+        """
+        Arguments:
+            - state (Overcooked_mdp.OvercookedState) object encoding the global view of the environment
+        returns:
+            - Normalized action probabilities determined by self.policy
+        """
+        # Preprocess the environment state
+        obs = self.featurize(state, debug=False)
+        my_obs = obs[self.agent_index]
+
+        # Compute non-normalized log probabilities from the underlying model
+        logits = self.policy.compute_actions(np.array([my_obs]), self.rnn_state)[2]['action_dist_inputs']
+
+        # Softmax in numpy to convert logits to normalized probabilities
+        return softmax(logits)
+
+    def action(self, state):
+        """
+        Arguments: 
+            - state (Overcooked_mdp.OvercookedState) object encoding the global view of the environment
+        returns: 
+            - the argmax action for a single observation state
+            - action_info (dict) that stores action probabilities under 'action_probs' key
+        """
+        # Preprocess the environment state
+        obs = self.featurize(state)
+        my_obs = obs[self.agent_index]
+
+        # Use Rllib.Policy class to compute action argmax and action probabilities
+        [action_idx], rnn_state, info = self.policy.compute_actions(np.array([my_obs]), self.rnn_state)
+        agent_action =  Action.INDEX_TO_ACTION[action_idx]
+        
+        # Softmax in numpy to convert logits to normalized probabilities
+        logits = info['action_dist_inputs']
+        action_probabilities = softmax(logits)
+
+        agent_action_info = {'action_probs' : action_probabilities}
+        self.rnn_state = rnn_state
+
+        return agent_action, agent_action_info
+
+
```

Comment:
I was a little confused by the need of the actions method. I didn't implement it because it's not used anywhere in the `evaluate_agent_pair` function (only time I use the `RllibAgent` class) and also because the convention I've normally seen is having a generic action function that operates on a batch dimension, so you could compute 10 actions by passing in a batch of size 10, or compute one single action by passing in a tensor of shape (1, ...). This is how I implemented `compute_actions` at the policy level, and was curious if you think it would be worth doing this at the agent level as well

Reply:
Ah, yeah this is a good point. I think the idea was to eventually switch `evaluate_agent_pair` to create multiple rollouts in parallel (rather than doing say 40 rollouts sequentially, one action at a time, requesting 40 actions at a time from each agent). This would enable to speed up agent evaluations quite dramatically when using NN-parameterized Agents. You'll see later that evaluations take a surprising amount of time, especially if you have a lot of different conditions to test, as you always want `n>40` (`n=100` is really what one should be doing, and even then sometimes there is high variance), so this is something you probably will want to improve in the future.

We were also using the `.actions` method already in the baselines code (to effectively do the same thing I was mentioning above, as the baselines runner was parallelized in the way I was describing), so in the previous version of the code agents would actually break without the `.actions` method, but you are right that it's currently not necessary.